internal/analysis/individual.go

```diff
@@ -19,221 +18,75 @@ import (
-		}
-
-		return AnalysisPoolResult{result, runTime}, err
-	})
-
+	err = job.Validate()
 	if err != nil {
-		return nil, fmt.Errorf("Failed to run analysis in a parallel pool: '%w'.", err)
-	}
-
-	// If the analysis was canceled, exit right away.
-	if poolResults.Canceled {
-		return nil, nil
-	}
-
-	poolResults.IsDone()
-
-	var errs error = nil
-	totalRunTime := int64(0)
-
-	for _, id := range fullSubmissionIDs {
-		err, ok := poolResults.WorkErrors[id]
-		if ok {
-			errs = errors.Join(errs, err)
-			continue
-		}
-
-		analysisPoolResult, ok := poolResults.Results[id]
-		if !ok {
-			errs = errors.Join(errs, fmt.Errorf("Unable to find result for submission: '%s'.", id))
-			continue
-		}
-
-		results[id] = analysisPoolResult.Result
-		totalRunTime += analysisPoolResult.RunTime
+		return nil, 0, nil, fmt.Errorf("Failed to validate job: '%w'.", err)
 	}
 
-	collectIndividualStats(fullSubmissionIDs, totalRunTime, options.InitiatorEmail)
-
-	return results, errs
-}
-
-func runSingleIndividualAnalysis(options AnalysisOptions, fullSubmissionID string) (*model.IndividualAnalysis, int64, error) {
-	// Lock this id so we don't try to do the analysis multiple times.
-	lockKey := fmt.Sprintf("analysis-individual-%s", fullSubmissionID)
-	lockmanager.Lock(lockKey)
-	defer lockmanager.Unlock(lockKey)
-
-	// The context has been canceled while waiting for a lock, abandon this analysis.
-	if options.Context.Err() != nil {
-		return nil, 0, nil
+	output := job.Run()
+	if output.Error != nil {
+		return nil, 0, nil, fmt.Errorf("Failed to run individual analysis job '%s': '%w'.", output.ID, output.Error)
 	}
 
-	startTime := timestamp.Now()
+	workErrors := make(map[string]string, len(output.WorkErrors))
 
-	// Check the DB for a complete analysis.
-	if !options.OverwriteCache {
-		result, err := db.GetSingleIndividualAnalysis(fullSubmissionID)
-		if err != nil {
-			return nil, 0, fmt.Errorf("Failed to check DB for cached individual analysis for '%s': '%w'.", fullSubmissionID, err)
-		}
+	if len(output.WorkErrors) != 0 {
+		for fullSubmissionID, err := range output.WorkErrors {
+			workErrors[fullSubmissionID] = err.Error()
 
-		if result != nil {
-			return result, 0, nil
-		}
-	}
+			logAttributes := submissionIDToLogValues(fullSubmissionID)
 
-	// Nothing cached, compute the analysis.
-	result, err := computeSingleIndividualAnalysis(options, fullSubmissionID, true)
-	if err != nil {
-		return nil, 0, fmt.Errorf("Failed to compute individual analysis for '%s': '%w'.", fullSubmissionID, err)
-	}
+			logAttributes = append([]any{err}, logAttributes...)
 
-	// Store the result.
-	if !options.DryRun && (options.Context.Err() == nil) {
-		err = db.StoreIndividualAnalysis([]*model.IndividualAnalysis{result})
-		if err != nil {
-			return nil, 0, fmt.Errorf("Failed to store individual analysis for '%s' in DB: '%w'.", fullSubmissionID, err)
+			log.Error("Failed to run individual analysis.", logAttributes...)
```

Comment:
Just a possible nit.

I would put the two `logAttributes` together (no newline between them).
You can also put the log.Error() in the same visual block.